File: backend/app/api/zotero.py
from __future__ import annotations

from fastapi import APIRouter, Response
from pydantic import BaseModel, Field

from app.zotero.models import ZoteroRelationType
from app.zotero.provider import get_zotero_provider
from app.zotero.store import ZoteroLinkStore, zotero_store_path

router = APIRouter(prefix="/zotero")


class BibtexExportRequest(BaseModel):
    item_keys: list[str] | None = None


class ZoteroLinkWrite(BaseModel):
    item_key: str = Field(min_length=1)
    relation_type: ZoteroRelationType = "related"
    note: str | None = None


def get_zotero_link_store() -> ZoteroLinkStore:
    return ZoteroLinkStore(zotero_store_path())


@router.get("/status")
def zotero_status() -> dict[str, object]:
    return get_zotero_provider().status().to_dict()


@router.get("/items")
def search_zotero_items(q: str = "", limit: int = 20) -> dict[str, object]:
    items = [item.to_dict() for item in get_zotero_provider().search(q, limit=max(1, min(limit, 100)))]
    return {"items": items, "total": len(items), "query": q}


@router.get("/items/{item_key}")
def get_zotero_item(item_key: str) -> dict[str, object]:
    item = get_zotero_provider().get_item(item_key)
    if item is None:
        return {"item": None}
    return item.to_dict()


@router.post("/export/bibtex")
def export_zotero_bibtex(request: BibtexExportRequest) -> dict[str, object]:
    bibtex = get_zotero_provider().export_bibtex(request.item_keys)
    return {"bibtex": bibtex, "item_count": len(request.item_keys or _entry_keys(bibtex))}


@router.get("/links/{article_id}")
def list_zotero_links(article_id: str) -> dict[str, object]:
    provider = get_zotero_provider()
    links = get_zotero_link_store().list_links(article_id)
    items = []
    for link in links:
        item = provider.get_item(link.zotero_item_key)
        items.append({"link": link.to_dict(), "item": item.to_dict() if item else None})
    return {"items": items, "total": len(items)}


@router.post("/links/{article_id}")
def create_zotero_link(article_id: str, request: ZoteroLinkWrite) -> dict[str, object]:
    link = get_zotero_link_store().upsert_link(
        article_id=article_id,
        zotero_item_key=request.item_key,
        relation_type=request.relation_type,
        note=request.note,
    )
    return link.to_dict()


@router.delete("/links/{article_id}/{item_key}", status_code=204)
def delete_zotero_link(article_id: str, item_key: str) -> Response:
    get_zotero_link_store().delete_link(article_id=article_id, zotero_item_key=item_key)
    return Response(status_code=204)


def _entry_keys(bibtex: str) -> list[str]:
    return [line for line in bibtex.splitlines() if line.startswith("@")]

File: frontend/src/lib/zotero.ts
export type ZoteroStatus = {
  provider: string;
  available: boolean;
  read_only: boolean;
  base_url?: string | null;
  version?: string | null;
  error?: string | null;
};

export type ZoteroItem = {
  item_key: string;
  bibtex_key: string;
  title: string;
  creators: string[];
  year: string | null;
  item_type: string;
  publication_title: string | null;
  doi: string | null;
  url: string | null;
  abstract_note: string | null;
  tags: string[];
  collections: string[];
  updated_at: string | null;
};

export type ZoteroItemSearchResponse = {
  items: ZoteroItem[];
  total: number;
  query: string;
};

export type ZoteroRelationType = "related" | "cites" | "background";

export type ZoteroArticleLink = {
  article_id: string;
  zotero_item_key: string;
  relation_type: ZoteroRelationType;
  created_at: string;
  note: string | null;
};

export type ZoteroArticleLinkItem = {
  link: ZoteroArticleLink;
  item: ZoteroItem | null;
};

export type ZoteroArticleLinksResponse = {
  items: ZoteroArticleLinkItem[];
  total: number;
};

export type BibtexExportResponse = {
  bibtex: string;
  item_count: number;
};

const API_BASE_URL = process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://localhost:8000";

export async function fetchZoteroStatus(): Promise<ZoteroStatus> {
  return requestJson<ZoteroStatus>("/zotero/status");
}

export async function searchZoteroItems(query: string, limit = 20): Promise<ZoteroItemSearchResponse> {
  const url = new URL("/zotero/items", API_BASE_URL);
  url.searchParams.set("q", query.trim());
  url.searchParams.set("limit", String(limit));
  return requestJsonUrl<ZoteroItemSearchResponse>(url);
}

export async function fetchArticleZoteroLinks(articleId: string): Promise<ZoteroArticleLinksResponse> {
  return requestJson<ZoteroArticleLinksResponse>(`/zotero/links/${articleId}`);
}

export async function createArticleZoteroLink(
  articleId: string,
  itemKey: string,
  relationType: ZoteroRelationType,
  note: string | null,
): Promise<ZoteroArticleLink> {
  return requestJson<ZoteroArticleLink>(`/zotero/links/${articleId}`, {
    body: JSON.stringify({ item_key: itemKey, relation_type: relationType, note }),
    method: "POST",
  });
}

export async function deleteArticleZoteroLink(articleId: string, itemKey: string): Promise<void> {
  const response = await fetch(new URL(`/zotero/links/${articleId}/${itemKey}`, API_BASE_URL).toString(), {
    cache: "no-store",
    method: "DELETE",
  });
  if (!response.ok) {
    throw new Error(`Zotero request failed: ${response.status}`);
  }
}

export async function exportZoteroBibtex(itemKeys: string[]): Promise<BibtexExportResponse> {
  return requestJson<BibtexExportResponse>("/zotero/export/bibtex", {
    body: JSON.stringify({ item_keys: itemKeys }),
    method: "POST",
  });
}

async function requestJson<T>(path: string, init: RequestInit = {}): Promise<T> {
  return requestJsonUrl<T>(new URL(path, API_BASE_URL), init);
}

async function requestJsonUrl<T>(url: URL, init: RequestInit = {}): Promise<T> {
  const response = await fetch(url.toString(), {
    ...init,
    cache: "no-store",
    headers: {
      "Content-Type": "application/json",
      ...init.headers,
    },
  });
  if (!response.ok) {
    throw new Error(`Zotero request failed: ${response.status}`);
  }
  return response.json() as Promise<T>;
}
